src/nmod_poly/divrem_xnmc.c

```diff
@@ -0,0 +1,285 @@
+
+    for (j = 0; j < r; j++)
+    {
+        val = nmod_mul(RQ[i+n+j], c, mod);
+        RQ[i+j] = n_addmod(val, A[i+j], mod.n);
+    }
+
+    i -= n;
+    while (i >= 0)
+    {
+        for (j = 0; j < n; j++)
+        {
+            val = nmod_mul(RQ[i+n+j], c, mod);
+            RQ[i+j] = n_addmod(val, A[i+j], mod.n);
+        }
+        i -= n;
+    }
+}
+
+/* division by x**n - c, with precomputation on c */
+/* constraint: modn < 2**(FLINT_BITS-1) */
+void _nmod_poly_divrem_xnmc_precomp(nn_ptr RQ, nn_srcptr A, slong len, ulong n, ulong c, ulong c_precomp, ulong modn)
+{
+    /* assumes len >= n */
+    slong i;
+    ulong j, r, val;
+
+    if (RQ != A)
+        for (j = 0; j < n; j++)
+            RQ[len-n+j] = A[len-n+j];
+
+    r = len % n;
+    i = len - r - n;  /* multiple of n, >= 0 by assumption */
+
+    for (j = 0; j < r; j++)
+    {
+        val = n_mulmod_shoup(c, RQ[i+n+j], c_precomp, modn);
+        RQ[i+j] = n_addmod(val, A[i+j], modn);
+    }
+
+    i -= n;
+    while (i >= 0)
+    {
+        for (j = 0; j < n; j++)
+        {
+            val = n_mulmod_shoup(c, RQ[i+n+j], c_precomp, modn);
+            RQ[i+j] = n_addmod(val, A[i+j], modn);
+        }
+        i -= n;
+    }
+}
+
+/* division by x**n - c, lazy with precomputation */
+/* constraint: max(A) + 2*modn <= 2**FLINT_BITS */
+/* coeff bounds: in [0, max(A)] | out [0, max(A) + 2*modn) */
+void _nmod_poly_divrem_xnmc_precomp_lazy(nn_ptr RQ, nn_srcptr A, slong len, ulong n, ulong c, ulong c_precomp, ulong modn)
+{
+    /* assumes len >= n */
+    slong i;
+    ulong j, r, val, p_hi, p_lo;
+
+    if (RQ != A)
+        for (j = 0; j < n; j++)
+            RQ[len-n+j] = A[len-n+j];
+
+    r = len % n;
+    i = len - r - n;  /* multiple of n, >= 0 by assumption */
+
+    for (j = 0; j < r; j++)
+    {
+        /* computes either val = (c*val mod n) or val = (c*val mod n) + n */
+        val = RQ[i+n+j];
+        umul_ppmm(p_hi, p_lo, c_precomp, val);
+        val = c * val - p_hi * modn;
+        /* lazy addition, yields RQ[i+j] in [0..k+2n), where max(RQ) <= k */
+        RQ[i+j] = val + A[i+j];
+    }
+
+    i -= n;
+    while (i >= 0)
+    {
+        for (j = 0; j < n; j++)
+        {
+            /* computes either val = (c*val mod n) or val = (c*val mod n) + n */
+            val = RQ[i+n+j];
+            umul_ppmm(p_hi, p_lo, c_precomp, val);
+            val = c * val - p_hi * modn;
+            /* lazy addition, yields RQ[i+j] in [0..k+2n), where max(RQ) <= k */
+            RQ[i+j] = val + A[i+j];
+        }
+        i -= n;
+    }
+}
+
+void nmod_poly_divrem_xnmc(nmod_poly_t Q, nmod_poly_t R, nmod_poly_t A, ulong n, ulong c)
+{
+    const ulong len = A->length;
+
+    if (len <= n)
+    {
+        nmod_poly_zero(Q);
+        nmod_poly_set(R, A);
+        return;
+    }
+
+    if (c == 0)
+    {
+        nmod_poly_set_trunc(R, A, n);
+        nmod_poly_shift_right(Q, A, n);
+        return;
+    }
+
+    int lazy = 0;
+    nn_ptr RQ = _nmod_vec_init(len);
```

Comment:
A temporary allocation would be good to try here.  @fredrik-johansson how do we deal with stack allocation for these types?